### ISSUE TITLE
Tests fail with latest Netty version

### ISSUE THREAD

**Issue by buchgr** [grpc/grpc-java#154]
When running gradle clean build on master, I get the following unit test failures.

```
:grpc-netty:processTestResources UP-TO-DATE
:grpc-netty:testClasses
:grpc-netty:test

io.grpc.transport.netty.NettyServerHandlerTest > sendFrameShouldSucceed FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:263

io.grpc.transport.netty.NettyServerHandlerTest > inboundDataWithEndStreamShouldForwardToStreamListener FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:263

io.grpc.transport.netty.NettyServerHandlerTest > closeShouldCloseChannel FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:246

io.grpc.transport.netty.NettyServerHandlerTest > clientCancelShouldForwardToStreamListener FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:263

io.grpc.transport.netty.NettyServerHandlerTest > connectionErrorShouldCloseChannel FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:263

io.grpc.transport.netty.NettyServerHandlerTest > streamErrorShouldNotCloseChannel FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:263

io.grpc.transport.netty.NettyServerHandlerTest > clientHalfCloseShouldForwardToStreamListener FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:263

io.grpc.transport.netty.NettyServerHandlerTest > inboundDataShouldForwardToStreamListener FAILED
    java.lang.NullPointerException at NettyServerHandlerTest.java:263

io.grpc.transport.netty.NettyClientHandlerTest > sendFrameShouldSucceed FAILED
    org.mockito.exceptions.verification.NeverWantedButInvoked at NettyClientHandlerTest.java:213

io.grpc.transport.netty.NettyClientHandlerTest > receivedGoAwayShouldFailQueuedStreams FAILED
    java.lang.NullPointerException at NettyClientHandlerTest.java:273

io.grpc.transport.netty.NettyClientHandlerTest > inboundHeadersShouldForwardToStream FAILED
    java.lang.NullPointerException at NettyClientHandlerTest.java:232

io.grpc.transport.netty.NettyClientHandlerTest > inboundDataShouldForwardToStream FAILED
    java.lang.NullPointerException at NettyClientHandlerTest.java:326

io.grpc.transport.netty.NettyClientHandlerTest > createStreamShouldSucceed FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at NettyClientHandlerTest.java:137

io.grpc.transport.netty.NettyClientHandlerTest > channelShutdownShouldFailInFlightStreams FAILED
    java.lang.NullPointerException at NettyClientHandlerTest.java:307

io.grpc.transport.netty.NettyClientHandlerTest > receivedGoAwayShouldFailUnknownStreams FAILED
    java.lang.NullPointerException at NettyClientHandlerTest.java:284
```

Example Stacktrace

```
channelShutdownShouldFailInFlightStreams

java.lang.NullPointerException
	at io.grpc.transport.netty.NettyClientHandler.onStreamError(NettyClientHandler.java:223)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.onException(Http2ConnectionHandler.java:284)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder$FlowControlledHeaders.error(DefaultHttp2ConnectionEncoder.java:501)
	at io.netty.handler.codec.http2.DefaultHttp2RemoteFlowController$FlowState$Frame.writeError(DefaultHttp2RemoteFlowController.java:543)
	at io.netty.handler.codec.http2.DefaultHttp2RemoteFlowController$FlowState.clear(DefaultHttp2RemoteFlowController.java:453)
	at io.netty.handler.codec.http2.DefaultHttp2RemoteFlowController$2.streamInactive(DefaultHttp2RemoteFlowController.java:79)
	at io.netty.handler.codec.http2.DefaultHttp2Connection.deactivateInternal(DefaultHttp2Connection.java:222)
	at io.netty.handler.codec.http2.DefaultHttp2Connection.access$700(DefaultHttp2Connection.java:52)
	at io.netty.handler.codec.http2.DefaultHttp2Connection$DefaultStream.close(DefaultHttp2Connection.java:402)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.closeStream(Http2ConnectionHandler.java:260)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.channelInactive(Http2ConnectionHandler.java:193)
	at io.grpc.transport.netty.NettyClientHandler.channelInactive(NettyClientHandler.java:204)
	at io.grpc.transport.netty.NettyClientHandlerTest.channelShutdownShouldFailInFlightStreams(NettyClientHandlerTest.java:307)
```

**Comment by buchgr** [comment]
I did some investigation and tried to build / run those tests on two different machines that have never seen a byte of grpc-java before. In both cases I got the exact same errors as on my workstation (additional computers where Linux and OSX, tried with Oracle Java 7u75 and 8u31).

Disclaimer: I have little to no experience with / knowledge of Mockito so there might be some magic that I am missing.

Looking at the first test that failed `NettyServerHandlerTest.sendFrameShouldSucceed()`.

```
java.lang.NullPointerException: executor
	at io.netty.util.concurrent.DefaultPromise.<init>(DefaultPromise.java:76)
	at io.netty.channel.DefaultChannelPromise.<init>(DefaultChannelPromise.java:50)
	at io.netty.handler.codec.http2.Http2CodecUtil$SimpleChannelPromiseAggregator.<init>(Http2CodecUtil.java:224)
	at io.netty.handler.codec.http2.DefaultHttp2FrameWriter.writeHeadersInternal(DefaultHttp2FrameWriter.java:386)
	at io.netty.handler.codec.http2.DefaultHttp2FrameWriter.writeHeaders(DefaultHttp2FrameWriter.java:157)
	at io.grpc.transport.netty.NettyHandlerTestBase.headersFrame(NettyHandlerTestBase.java:92)
	at io.grpc.transport.netty.NettyServerHandlerTest.createStream(NettyServerHandlerTest.java:263)
	at io.grpc.transport.netty.NettyServerHandlerTest.sendFrameShouldSucceed(NettyServerHandlerTest.java:142)
```

The methods where it starts to get interesting are

```java
  protected final ByteBuf headersFrame(int streamId, Http2Headers headers) {
    ChannelHandlerContext ctx = newContext();
    frameWriter.writeHeaders(ctx, streamId, headers, 0, false, promise);
    return captureWrite(ctx);
  }
```

and subsequently

```java
  protected final ChannelHandlerContext newContext() {
    ChannelHandlerContext ctx = Mockito.mock(ChannelHandlerContext.class);
    when(ctx.alloc()).thenReturn(UnpooledByteBufAllocator.DEFAULT);
    return ctx;
  }
```

Note that on the `ChannelHandlerContext` created in `newContext()` only the `ctx.alloc()` method gets stubbed. No stub for `ctx.executor()`, `ctx.channel()` and so on. So it seems to make sense (and like a bug) that subsequently in `frameWriter.writeHeaders`  the call to `ctx.executor()` returns `null` and leads to the reported NPE when trying to create a `DefaultChannelPromise` object ...

So when I now add some stubs for `ctx.executor()` and `ctx.channel()` the call to `frameWriter.writeHeaders(...)` works fine without any errors. However, in the subsequent call to `captureWrite(ctx)` I get a Mockito error saying `ctx.write(...) was called more than once ....`

The implementation of `captureWrite(...)`, besides capturing the argument, also seems to try to verify that `ctx.write(...)` is called exactly once

```java
  protected final ByteBuf captureWrite(ChannelHandlerContext ctx) {
    ArgumentCaptor<ByteBuf> captor = ArgumentCaptor.forClass(ByteBuf.class);
    verify(ctx).write(captor.capture(), any(ChannelPromise.class));
    return captor.getValue();
  }
```

Now looking at the implementation of `frameWriter.writeHeaders(...)` there are not one, but at least TWO calls to `ctx.write(...)` [1]. So how can this test be correct?

There is lots of other examples in these tests that on first sight seem to be broken (to me), but then again the tests seem to work fine at least on @ejona86 workstation, so I am thinking there is probably some magic involved that I am missing @zhangkun83 @ejona86 ? Are those tests correct?

[1] https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java#L423

**Comment by ejona86** [comment]
@buchgr, could it be you are using a newer version of Netty than the one in lib/netty submodule?

**Comment by buchgr** [comment]
@ejona86 *face palm* ... yes you are right. due to my recent netty work I overwrote my local maven repo. works with the submodule. sorry. so stupid. :P (but now the integration tests no longer pass haha).

**Comment by buchgr** [comment]
Hmm ... but on a different note, we should probably not rely on Netty's internal implementation details (i.e. how often a function gets called, or whether the `EventLoop` is accessed via `ChannelHandlerContext.executor()` or `Channel.eventLoop()`) in our unit tests?

**Comment by nmittler** [comment]
I'm not sure if we've updated Netty since the flow control changes by @louiscryan.  I seem to recall those changes causing some of our tests to fail, but I'm not sure anyone has looked into it.

**Comment by ejona86** [comment]
Netty recently released 5.0.0.Alpha2 with Maven artifacts. If we get the tests working, then we can depend on it for a 0.5.0 grpc-java release.

**Comment by buchgr** [comment]
Nathan suggested to "downgrade" to 4.1. I think it makes sense because it's still at least one year before Netty 5 is out and we aren't using any of the Netty 5 features anyway.

**Comment by ejona86** [comment]
4.1 is fine, too, but it is alpha as well and thus basically the same in my mind. If it is truly no more work to do 4.1 instead of 5, 4.1 works. I'd mainly be concerned that going to 4.1 turns into a rabbit hole because of a bigger delta in changes and it consumes considerably more time.

**Comment by hepin1989** [comment]
I think 4.1.0 beta4 is fine,but don't using epoll trasnport,there is a bug in beta4

**Comment by nmittler** [comment]
Fixed by #179.
